### PR TITLE
Spring Boot 3

### PR DESCRIPTION
**Pull request by shubham1g5**
## Technical Summary

[JIRA](https://dimagi.atlassian.net/browse/USH-4496)

Update to [Spring Boot 3](https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-3.0-Migration-Guide)

### Safety story

- Relying on automated unit tests and staging scripts
- Will keep on staging for a week or so to catch anything unexpected

### QA Plan

https://dimagi.atlassian.net/browse/QA-7246

### Special deploy instructions

- [x] This PR can be deployed after merge with no further considerations.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations.

### Review

- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change.

### CODE REVIEW

File: src/main/java/org/commcare/formplayer/configuration/WebSecurityConfig.java [hunk 1]
```diff
@@ -70,11 +75,18 @@ public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
...
         http.addFilterAfter(getSessionAuthFilter(authenticationManager), HmacAuthFilter.class);
-        http.csrf()
-                .csrfTokenRepository(CookieCsrfTokenRepository.withHttpOnlyFalse())
-                .ignoringRequestMatchers(new RequestHeaderRequestMatcher(Constants.HMAC_HEADER));
-        http.cors();
 
+        // By setting the csrfRequestAttributeName to null, the CsrfToken must first be loaded to determine what
+        // attribute name to use. This causes the CsrfToken to be loaded on every request.
+        CsrfTokenRequestAttributeHandler requestHandler = new CsrfTokenRequestAttributeHandler();
+        requestHandler.setCsrfRequestAttributeName(null);
```
Comment (shubham1g5) [inline src/main/java/org/commcare/formplayer/configuration/WebSecurityConfig.java:82]: Noting that this change was in response to a change made in Spring 3 that [defers the csrf tokens lookup](https://docs.spring.io/spring-security/reference/5.8/migration/servlet/exploits.html#_defer_loading_csrftoken) until needed optimising reads of `HttpSession`. Although that breaks all requests to FP as 403 error due to csrf token coming as `undefined` in request headers. In response, I first implemented the opt out steps mentioned [here](https://docs.spring.io/spring-security/reference/5.8/migration/servlet/exploits.html#_i_am_using_a_single_page_application_with_cookiecsrftokenrepository), that didn't solve the problem completely (Half the times requests were with csrf `undefined` and got errored with 403 while on trying again the csrf was there in request payloads, but I could not get to the bottom of why it was happening).  Ultimately I decided to completely opt out of deferred tokens behaviour by follow the steps listed [here](https://docs.spring.io/spring-security/reference/5.8/migration/servlet/exploits.html#_i_need_to_opt_out_of_deferred_tokens_for_another_reason).

File: src/main/resources/application.properties [hunk 2]
```diff
@@ -26,7 +26,6 @@ management.health.diskspace.threshold=1073742000
 
 # metrics
 management.statsd.metrics.export.flavor=datadog
-management.metrics.web.server.request.autotime.enabled=false
```
Comment (Jtang-1) [inline src/main/resources/application.properties:29]: the migration guide mentions "Should be applied at the ObservationRegistry level.". Do we need to do that?

**Review by Jtang-1 — APPROVED** [review]

File: src/main/resources/application.properties (hunk 2 above)
Comment (shubham1g5) [reply src/main/resources/application.properties:29]: Do you have a reference to this ? I just saw that it's been removed.

Comment (Jtang-1) [reply src/main/resources/application.properties:29]: https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-3.0.0-RC1-Configuration-Changelog#removed-in-300-rc1

Comment (snopoke) [reply src/main/resources/application.properties:29]: More info here: https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-3.0-Migration-Guide#micrometer-and-metrics-changes and here: https://github.com/spring-projects/spring-boot/issues/41745

Overall I think this is OK though I don't recall why we turned that metric off.

File: src/test/java/org/commcare/formplayer/repo/FormSessionRepoTest.java [hunk 3]
```diff
@@ -64,17 +64,19 @@ public void testSaveAndLoad() {
...
         assertThat(dateCreated).isNotNull();
-        assertThat(loaded.getVersion()).isEqualTo(1);
+        assertThat(loaded.getVersion()).isEqualTo(0);
 
+        // Make a edit to trigger version update
+        loaded.setInitLang("newlang");
         formSessionRepo.saveAndFlush(loaded);
         assertThat(loaded.getDateCreated()).isEqualTo(dateCreated);
-        assertThat(loaded.getVersion()).isEqualTo(2);
+        assertThat(loaded.getVersion()).isEqualTo(1);
```
Comment (snopoke) [inline src/test/java/org/commcare/formplayer/repo/FormSessionRepoTest.java:79]: Can you explain these changes. They don't seem related to the Spring upgrade.

**Review by snopoke — COMMENTED** [review]
Overall this looks good to me.

File: src/test/java/org/commcare/formplayer/repo/FormSessionRepoTest.java (hunk 3 above)
Comment (shubham1g5) [reply src/test/java/org/commcare/formplayer/repo/FormSessionRepoTest.java:79]: I could not really find the related change documentation for this in hibernate but after upgrading to SB 3, the version doesn't increment unless you make a change to the object. My guess is that the change is caused by shifting to the Jakarta persistence apis from javax but I don't really have a supportive reference here that can explain this change.

**Comment by shubham1g5** [comment]
This has been on staging for a week and I have not seen anything related, as such I am planning to merge this after tomorrow's deploy so that it will get to prod in next week's deploy.

**Review by snopoke — APPROVED** [review]

**Comment by shubham1g5** [comment]
> Has someone tested SMS forms?

Not really, are you concerned because of the security changes ?

**Comment by snopoke** [comment]
> > Has someone tested SMS forms?
> 
> Not really, are you concerned because of the security changes ?

yes, that's one place where auth differs from webapps

**Comment by shubham1g5** [comment]
Great to know, I will make sure we get that tested. Thanks.

**Pull request merged**